src/zocalo/service/dispatcher.py

```diff
@@ -0,0 +1,217 @@
+        # be used to determine unique file paths.
+        recipe_id = parameters.get("uuid") or str(uuid.uuid4())
+        parameters["uuid"] = recipe_id
+
+        if rw:
+            # If we received a recipe wrapper then we already have a recipe_ID
+            # attached to logs. Make a note of the downstream recipe ID so that
+            # we can track execution beyond recipe boundaries.
+            self.log.info(
+                "Processing request with new recipe ID %s:\n%s", recipe_id, str(message)
+            )
+
+        # From here on add the global ID to all log messages
+        with self.extend_log("recipe_ID", recipe_id):
+            self.log.debug("Received processing request:\n" + str(message))
+            self.log.debug("Received processing parameters:\n" + str(parameters))
+
+            filtered_message = copy.deepcopy(message)
+            filtered_parameters = copy.deepcopy(parameters)
+
+            # Call a hook before processing the message
+            if not self.hook_before_filtering(header, message, parameters, recipe_id):
+                return
+
+            # Create empty recipe
+            filtered_message["recipe"] = workflows.recipe.Recipe()
+
+            # Apply all specified filters in order to message and parameters
+            message_filters = [
+                f.load()
+                for f in pkg_resources.iter_entry_points("zocalo.dispatcher.filters")
+            ] + [self.filter_load_recipes_from_files, self.filter_apply_parameters]
+
+            for f in message_filters:
+                try:
+                    filtered_message, filtered_parameters = f(
+                        filtered_message, filtered_parameters
+                    )
+                except Exception as e:
+                    self.log.error(
+                        "Rejected message due to filter error: %s",
+                        str(e),
+                        exc_info=True,
+                    )
+                    self._transport.nack(header)
+                    return
+
+            self.log.debug("Mangled processing request:\n" + pformat(filtered_message))
+            self.log.debug(
+                "Mangled processing parameters:\n" + pformat(filtered_parameters)
+            )
+
+            # Create the RecipeWrapper object
+            rw = workflows.recipe.RecipeWrapper(
+                environment={"ID": recipe_id},
+                recipe=filtered_message["recipe"],
+                transport=self._transport,
+            )
+
+            # Conditionally acknowledge receipt of the message
+            txn = self._transport.transaction_begin()
+            self._transport.ack(header, transaction=txn)
+
+            # Call another hook just before dispatching the message
+            if not self.hook_before_dispatch(
+                header, message, parameters, recipe_id, filtered_message, rw,
+            ):
+                return
```

Comment:
This now leaves a dangling transaction.

I am also not 100% sure I understand what the hooks are supposed to achieve.

Reply:
Me neither, dont know what the use case was for these hooks (c/p from Tim's code), i assumed this had gone through you already, but maybe not. You discussed some of this with him here:
https://github.com/DiamondLightSource/python-zocalo/pull/14

Reply:
Got it. Needs moving above the transaction then.